src/otx/core/data/dataset/base.py

```diff
@@ -27,6 +29,28 @@
 Transforms = Union[Compose, Callable, List[Callable]]
 
 
+@dataclass
+class SubsetDataMetaInfo:
+    """Meta information of each subset datasets."""
+
+    class_names: list[str]
+
+    @property
+    def num_classes(self) -> int:
+        """Return number of classes."""
+        return len(self.class_names)
```

Comment:
What's the reason to introduce this?
Is there any use-case where a dataset can have different `num_classes` across its subsets?

Reply:
In the current status, I think there are no cases that you mentioned. I introduced this class as reasons below
1. At the H-label Classification, HLabelInfo should be generated at the `OTXDataset` level, not `OTXDataModule` level. So, I need to insert the `MetaInfo` from `OTXDataset` to `OTXDataModule`. In this situation, we need to construct the dictionary based `MetaInfo` since `OTXDataset` is the subset of datasets. Since current `OTXLitModule` communicates with the `OTXDataModule`, the overall flow is `OTXDataset generates the subset metainfo` --> `deliver to OTXDataModule` --> `deliver to OTXLitModule`.
2. I think some users might want to use different classes at the train and val, however, AFAIK, it is prohibited now from Datumaro.import_from. So, I want to make additional room for this case. 
3. I'm not 100% sure about the case, however, maybe Self-SL scenario looks similar with number 2.

Reply:
> 2. I think some users might want to use different classes at the train and val, however, AFAIK, it is prohibited now from Datumaro.import_from. So, I want to make additional room for this case.
> 3. I'm not 100% sure about the case, however, maybe Self-SL scenario looks similar with number 2.

I don't think so. Rather, if there is a mismatch in labels between subsets, we should notify to the user with an appropriate exception to fail as fast as it can. In addition, in the current flow, it is impossible that they are different between subsets. AFAIK, it is always `dm_dataset.subsets().categories == dm_dataset.categories`.

> At the H-label Classification, HLabelInfo should be generated at the OTXDataset level, not OTXDataModule level.

How `HLabelInfo` can be different between subsets unless `dm_dataset.subsets().categories == dm_dataset.categories`? Even if they are different, so that `OTXDataModule` can have different `HLabelInfo` as its member variables, in that case, you should raise an exception to notify it to an user.

Reply:
> I don't think so. Rather, if there is a mismatch in labels between subsets, we should notify to the user with an appropriate exception to fail as fast as it can. In addition, in the current flow, it is impossible that they are different between subsets. AFAIK, it is always dm_dataset.subsets().categories == dm_dataset.categories.

How about Self-SL case ? In that case, the label of training data and validation data will be different. I know it's extreme case, however, we need to consider. 

> How HLabelInfo can be different between subsets unless dm_dataset.subsets().categories == dm_dataset.categories? Even if they are different, so that OTXDataModule can have different HLabelInfo as its member variables, in that case, you should raise an exception to notify it to an user.

What I said is not about the label difference. To make H-label annotation, Datumaro label should be converted to the h-label format label. (https://github.com/sungmanc/training_extensions/blob/d2fd2089fe44c8c81a573760634ea0992750fa48/src/otx/core/data/dataset/classification.py#L142) and `HLabelInfo` is needed in this phase. So, if we follow the current implementation, we might need the code below.
```python3
        dataset = DmDataset.import_from(self.config.data_root, format=self.config.data_format)

        config_mapping = {
            self.config.train_subset.subset_name: self.config.train_subset,
            self.config.val_subset.subset_name: self.config.val_subset,
            self.config.test_subset.subset_name: self.config.test_subset,
        }

        mem_size = parse_mem_cache_size_to_int(config.mem_cache_size)
        mem_cache_mode = (
            "singleprocessing"
            if all(config.num_workers == 0 for config in config_mapping.values())
            else "multiprocessing"
        )
        mem_cache_handler = MemCacheHandlerSingleton.create(
            mode=mem_cache_mode,
            mem_size=mem_size,
        )

        meta_info_dict: dict[str, SubsetDataMetaInfo] = {}
        for name, dm_subset in dataset.subsets().items():
            if name not in config_mapping:
                log.warning(f"{name} is not available. Skip it")
                continue

            self.subsets[name] = OTXDatasetFactory.create(
                task=self.task,
                dm_subset=dm_subset,
                mem_cache_handler=mem_cache_handler,
                cfg_subset=config_mapping[name],
                cfg_data_module=config,
            )
            
            ### If `MetaInfo` is generated from the `OTXDataModule` level,
            ### Need to check the task type to deliver the MetaInfo to the `OTXDataset`
            ### I want to avoid this if/else sentence to avoid the complexity.
            ### That's the reason why I introduced the `SubsetMetaInfo`
            
            if self.task == OTXTaskType.H_LABEL_CLS:
                self.subsets[name].meta_info = self.meta_info
```

If the code flow above looks okay or have other design, please let me know. I'm not a big fan of `SubsetMetaInfo`.

Reply:
As I said, we have to exclude that case and notify it to user if there is a mismatch. Therefore,

```python
        meta_infos = []
        for name, dm_subset in dataset.subsets().items():
            ...
            meta_infos += [self.subsets[name].meta_info]
        
        if has_mismatch(meta_infos):
            msg = "..."
            raise ValueError(msg)
         self.meta_info = next(iter(meta_infos))
```

Reply:
I got your points. https://github.com/openvinotoolkit/training_extensions/pull/2761/commits/a0f493ce8502ca45645315da89367d2a48879017